.github/workflows/e2e-external.yml

```diff
@@ -0,0 +1,53 @@
+name: End-to-end tests for external PRs
+
+on:
+  pull_request_review:
+    types: [submitted]
+
+concurrency:
+  group: ${{ github.event_name }}-${{ github.ref }}
+
+jobs:
+  test:
+    name: Run end to end tests
+    # If reviewed by a repo(/org) owner
+    if: |
+      github.event.pull_request.author_association != 'MEMBER'
+      && github.event.review.author_association == 'MEMBER'
+      && github.event.review.state == 'approved'
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          # Important: use the commit that was reviewed. GitHub is making sure
+          # that this is race-condition-proof
+          ref: ${{ github.event.review.commit_id }}
+
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
+
+      - name: Install Poetry
+        run: |
+          pipx install poetry --python=python3.10
+
+      - name: Poetry caches
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cache/
+          key: ${{ hashFiles('poetry.lock') }}
+
+      - name: Install deps
+        run: poetry install
+
+      - name: Run end-to-end tests
+        run: poetry run pytest tests/end_to_end
+        env:
+          COVERAGE_COMMENT_E2E_GITHUB_TOKEN_USER_1: ${{ secrets.COVERAGE_COMMENT_E2E_GITHUB_TOKEN_USER_1 }}
+          COVERAGE_COMMENT_E2E_GITHUB_TOKEN_USER_2: ${{ secrets.COVERAGE_COMMENT_E2E_GITHUB_TOKEN_USER_2 }}
+          COVERAGE_COMMENT_E2E_ACTION_REF: ${{ github.event.review.commit_id }}
+          COVERAGE_COMMENT_E2E_REPO_SUFFIX: ${{ github.event.pull_request.number }}
+          COVERAGE_COMMENT_E2E_PR_AUTHOR: ${{ github.event.pull_request.user.login }}
```

Comment:
```suggestion
      - name: Check cache if the tests have already run
        id: cache
        uses: actions/cache/restore@v3
        with:
          key: ${{ github.ref }}-${{ github.event.review.commit_id }}
          path: dummy.txt

      - name: Checkout
        if: steps.cache.outputs.cache-hit != 'true'
        uses: actions/checkout@v3
        with:
          # Important: use the commit that was reviewed. GitHub is making sure
          # that this is race-condition-proof
          ref: ${{ github.event.review.commit_id }}

      - name: Set up Python
        if: steps.cache.outputs.cache-hit != 'true'
        uses: actions/setup-python@v4
        with:
          python-version: "3.10"

      - name: Install Poetry
        if: steps.cache.outputs.cache-hit != 'true'
        run: |
          pipx install poetry --python=python3.10

      - name: Poetry caches
        if: steps.cache.outputs.cache-hit != 'true'
        uses: actions/cache@v3
        with:
          path: |
            ~/.cache/
          key: ${{ hashFiles('poetry.lock') }}

      - name: Install deps
        if: steps.cache.outputs.cache-hit != 'true'
        run: poetry install

      - name: Run end-to-end tests
        if: steps.cache.outputs.cache-hit != 'true'
        run: poetry run pytest tests/end_to_end
        env:
          COVERAGE_COMMENT_E2E_GITHUB_TOKEN_USER_1: ${{ secrets.COVERAGE_COMMENT_E2E_GITHUB_TOKEN_USER_1 }}
          COVERAGE_COMMENT_E2E_GITHUB_TOKEN_USER_2: ${{ secrets.COVERAGE_COMMENT_E2E_GITHUB_TOKEN_USER_2 }}
          COVERAGE_COMMENT_E2E_ACTION_REF: ${{ github.event.review.commit_id }}
          COVERAGE_COMMENT_E2E_REPO_SUFFIX: ${{ github.event.pull_request.number }}
          COVERAGE_COMMENT_E2E_PR_AUTHOR: ${{ github.event.pull_request.user.login }}

      - name: Create dummy file
        if: steps.cache.outputs.cache-hit != 'true'
        run: touch dummy.txt
        shell: bash

      - name: Create cache to indicate that the tests have run
        if: steps.cache.outputs.cache-hit != 'true'
        uses: actions/cache/save@v3
        with:
          key: ${{ github.ref }}-${{ github.event.review.commit_id }}
          path: dummy.txt
```

> I see one remaining problem that I don't know how to solve and if it's worth solving: if we both approve a PR, e2e tests will run twice.

I think the easiest way to make this work is to use the cache system and to create a cache for the commit hash when the tests run. Whether it's worth it I don't know either, but at least it wouldn't add so much complexity.

Reply:
I could be wrong, but I think if:
- you approve a PR
- tests are red
- then I approve the PR

Then the step is going to run with everything skipped, so it will succeed, and turn the check green (honnestly, I'm not even entirely sure that there will be a check displayed on the UI. I'd like to say there will but we'll need to test this)

Reply:
> Then the step is going to run with everything skipped

If one step of the workflow fails, won't all subsequent steps be canceled? That was actually my reasoning behind this. Since the cache is added in the very last step, this should only happen if all steps before (i.e. also the tests) were successful.

Reply:
Ok, I've tested and:
- The result of the workflow _do_ appear in the PR checks
- But the last run overrides the state, whether this be a success, a failure of even the whole step skipped with an if

I'm going to add something easier: e2e tests will only launch if the review body contains /e2e

Reply:
> I'm going to add something easier: e2e tests will only launch if the review body contains /e2e

Sure, that works too